### PR TITLE
chore(weave): Simple pagination

### PR DESCRIPTION
**Pull request by jwlee64**
edits use calls to batch requests into pages of 10k instead of one big request
it does sequential requests of 10k up to 50 times in a row
so we can send more data over the wire without actually implementing pagination

still accounts for passing a limit of 1 in to check if calls exist

current load times for 300k calls

single call = 37s total load time

Screenshot 2024-03-13 at 4 44 36 PM

30 calls of 10k pages = 40s total load time

Screenshot 2024-03-13 at 4 43 43 PM

### CODE REVIEW

**Review by tssweeney — CHANGES_REQUESTED** [review]
We want to keep the interface as tight and small as possible. Since this pagination param (page size) is purely in implementation/performance param, we should not expose it to the caller. Instead, the `useCalls` implementation itself should internally use this paginated approach at all times. This hides the implementation details of how all the calls are fetched from the calling code and allows us to refactor more easily.

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts [hunk 1]
```diff
@@ -98,43 +98,96 @@ const useCalls = (
...
+  const [allCallsLoaded, setAllCallsLoaded] = useState(opts?.skip ?? false);
+  const [callRes, setCallRes] = useState<traceServerClient.TraceCallSchema[]>(
+    []
+  );
   const deepFilter = useDeepMemo(filter);
+
+  // if a limit is provided, we use that and only return the single page
+  // otherwise we use a default of 10,000 and load 50 pages to return all calls
+  // Arbitrary page limit of 10,000, in conjunction with the max 50 pages limits calls to 500,000
+  const pageSize = limit ?? 10000;
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:112]: please make the 10000 and 50 constants at the top of the file

Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:112]: I think you might have an issue if the user passes a limit > 10_000

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts [hunk 2]
```diff
@@ -98,43 +98,96 @@ const useCalls = (
...
+            call_ids: deepFilter.callIds,
+            trace_roots_only: deepFilter.traceRootsOnly,
+            wb_run_ids: deepFilter.runIds,
+            wb_user_ids: deepFilter.userIds,
+          },
+          limit: pageSize,
+          offset: pageNumber * pageSize,
+        });
+
+        if (res.calls.length < pageSize || pageNumber + 1 > maxPages) {
```
Comment (tssweeney) [inline weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:143]: nit: `+ 1 >` is harder to understand than `==`

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts (hunk 1 above)
Comment (jwlee64) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts:112]: I just figured if they really wanted a limit of over 10k , they can get it.
Would i be better to have some mechanism to get the exact amount of the limit if limit is over 10k?
eg limit of 47k
get 10 + 10 + 10 + 10 + 7?

**Review by tssweeney — APPROVED** [review]
Approving with Feedback

File: weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts (hunk 1 above)
Comment (jwlee64) [reply weave-js/src/components/PagePanelComponents/Home/Browse3/pages/wfReactInterface/tsDataModelHooks.ts]: gotcha i got that working

https://github.com/wandb/weave/assets/25037002/b98e0802-3e98-4d69-85fe-175cbcf7ace0

**Pull request merged**